packages/toolpad-app/src/toolpad/DomLoader.tsx

```diff
@@ -147,6 +159,55 @@ export function domLoaderReducer(state: DomLoader, action: DomAction): DomLoader
   }
 
   switch (action.type) {
+    case 'DOM_UPDATE_HISTORY': {
+      const updatedUndoStack = [...state.undoStack, state.pendingHistoryDom];
```

Comment:
Not sure I understand why we need `pendingHistoryDom`, can't we just use `state.dom` here?

Reply:
That was my first idea, I did try using `state.dom` here, but by the time this is called `state.dom` has updated values, where what I want is values that were in the dom before the actions were called. Storing a reference to last state that hasn't been pushed to history stack resolved this issue.
Just to be 100% sure I tried passing `state.dom` here after all these changes with throttle/debounce, but it still doesn't work.
Basically if we store `state.dom` here, we loose 1 point of history, and by using reference to "pending" state it always works as expected

Reply:
oh ok I see, it just means that the top of the stack contains the current state, and that doing an undo means first popping an item, and then resetting the dom to what's in the top of the stack instead of what we just popped off.

Reply:
Hmm, I'm not sure I fully understand whether you meant that this is how it's working now or how it would work if we pushed `state.dom` instead of `pendingHistoryDom` but the idea is to keep undo stack populated only with previous states 🤔 
Do you see a way that we could avoid using `pendingHistoryDom` ref? I was trying various things, like passing state to dispatch, etc... but none of them worked predictably and this (current) approach seemed to do the trick

Reply:
> Do you see a way that we could avoid using `pendingHistoryDom` ref?

I think it can be done if we slightly alter this philosophy: "the idea is to keep undo stack populated only with previous states" to "the idea is to keep undo stack populated with full history up until the current state". And initialize the `undoStack` with `[dom]` instead of `[]`. And then when popping an item off the stack, not use that popped off item to restore the dom, but look at the top item of the trimmed stack. 

Basically it would work the same way, only instead of adding an extra property for `pendingHistoryDom`, the top item in the stack would be `pendingHistoryDom`. I think there are some small benefits to this method:

* One less property to keep in sync `pendingHistoryDom`
* We will want to give an overview of the history of changes at some point. To do that we will have to add a human readable label to each entry in the stack. In the implementation with `pendingHistoryDom` there would be an entry missing for the last change, and we'd need to special case the last change. In the implementation without `pendingHistoryDom`, the `undoStack` would reflect the whole history up until the last change, and we can just flat render this array.

Then again, I may be wrong in my thinking, it's only based on reading the implementation and thinking about the consequences.

Reply:
@Janpot you are correct, thanks, another great tip 💪 I changed the implementation according to your suggestion. I guess the main con of this new approach is that undo/redo methods work slightly differently, but that should be fine 🤔 

> In the implementation with pendingHistoryDom there would be an entry missing for the last change, and we'd need to special case the last change

I don't think that's true - in the case of `pendingHistoryDom` initially we would have simply empty array, which means there is no history to undo. At the time we would trigger action and push `pendingHistoryDom` we could add a label. IMO specifically right now we will have to add extra checks if we want to render history and omit top most entry, i.e. ```
if (undoStack.length < 2) {
  // means there is no history to undo
}
```

We just need to remember in the future that top most item in the undo stack represents current state 🤔